tests/optimize/test_qnspsa.py

```diff
@@ -0,0 +1,236 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Unit tests for the ``QNSPSAOptimizer``
+"""
+import pytest
+
+import pennylane as qml
+from pennylane import numpy as np
+
+
+class TestSPSAOptimizer:
+    @pytest.mark.parametrize("seed", [1, 151, 1231])
+    @pytest.mark.parametrize("finite_diff_step", [1e-3, 1e-2, 1e-1])
+    def test_qnspsa_grad(self, finite_diff_step, seed):
+        """Test that the QNSPSA gradient estimation is correct by comparing the optimizer result
+        to a straightforward implementation."""
+        opt = qml.QNSPSAOptimizer(
+            stepsize=1e-3,
+            regularization=1e-3,
+            finite_diff_step=finite_diff_step,
+            resamplings=1,
+            blocking=True,
+            history_length=5,
+            seed=seed,
+        )
+        qnode, params_shape = self.get_single_input_qnode()
+        params = np.random.rand(*params_shape)
+
+        # gradient result from QNSPSAOptimizer
+        grad_tapes, grad_dirs = opt._get_spsa_grad_tapes(qnode, [params], {})
+        raw_results = qml.execute(grad_tapes, qnode.device, None)
+        grad_res = opt._post_process_grad(raw_results, grad_dirs)[0]
+
+        # gradient expected from a straightforward implementation
+        rng = np.random.default_rng(seed)
+        direction = rng.choice([-1, 1], size=params.shape)
+        loss_plus = qnode(params + direction * finite_diff_step)
+        loss_minus = qnode(params - direction * finite_diff_step)
+        grad_expected = (loss_plus - loss_minus) / (2 * finite_diff_step) * direction
```

Comment:
Trying to understand what does **analytically** mean here. Would you suggest building the unbiased non-stochastic version of the gradient and metric tensor estimation, and do unit tests on if the stochastic values finally converge around the analytical values when averaged over multiple trials?

Reply:
I mean deriving an expression for the gradient to directly use as the exact expected value. Since the gradient directions are stochastic, what you could do is first obtain the directions using `opt._get_spsa_grad_tapes`, then from that calculate what the gradient *should exactly be*.

Reply:
Still a bit confused on what would be different from the current L46-51. Other than getting the direction from ``opt._get_spsa_grad_tapes`` instead from the RNG api, is there another component that should be changed from L49-L51? Is the idea to use parameter shift rule instead of the finite difference for the unit test?

Reply:
So, say for example that `qnode` is the following:
```python
@qml.qnode(dev)
def qnode(x):
    qml.RY(x, wires=0)
    return qml.expval(qml.PauliZ(0))
```
Then instead of writing these two lines:
```python
loss_plus = qnode(params + direction * finite_diff_step)
loss_minus = qnode(params - direction * finite_diff_step)
grad_expected = (loss_plus - loss_minus) / (2 * finite_diff_step) * direction
```
You could just write
```python
grad_expected = np.sin(params) * np.sin(direction * finite_diff_step) * -direction / finite_diff_step
```
Note that I used the output of the QNode to calculate what the gradient should be. This is what I mean when I say derive an expression for the gradient, rather than copying the functionality of the optimizer.

The same applies to the test immediately below this. Deriving an expression for the metric tensor by hand is going to ensure that your optimizer functionality works as you would expect.